ARCs/arc-0056.md

```diff
@@ -0,0 +1,381 @@
+    boxes?: {
+      /** The app ID for the box */
+      app?: number;
+      /** The base64 encoded box key */
+      key: string;
+      /** The number of bytes being read from the box */
+      readBytes: number;
+      /** The number of bytes being written to the box */
+      writeBytes: number;
+    };
+    /** Recommended foreign accounts */
+    accounts?: string[];
+    /** Recommended foreign apps */
+    apps?: number[];
+    /** Recommended foreign assets */
+    assets?: number[];
+  };
+}
+```
+
+### Event Interface
+
+[ARC-28](./arc-0028.md) events are described using an extension of the original interface described in the ARC, with the addition of an optional struct field for arguments
+
+```ts
+interface Event {
+  /** The name of the event */
+  name: string;
+  /** Optional, user-friendly description for the event */
+  desc?: string;
+  /** The arguments of the event, in order */
+  args: Array<{
+    /** The type of the argument */
+    type: ABIType;
+    /** Optional, user-friendly name for the argument */
+    name?: string;
+    /** Optional, user-friendly description for the argument */
+    desc?: string;
+    /** If the type is a struct, the name of the struct */
+    struct?: StructName;
+  }>;
+}
+```
+
+### Type Interfaces
+
+The types defined in [ARC-4](./arc-0004.md) may not fully described the best way to use the ABI values as intended by the contract developers. These type interfaces are intended to supplement ABI types so clients can interact with the contract as intended.
+
+```ts
+/** An ABI-encoded type */
+type ABIType = string;
+
+/** The name of a defined struct */
+type StructName = string;
+
+/** Raw byteslice without the length prefixed that is specified in ARC-4 */
+type AVMBytes = "bytes";
+
+/** Mapping of named structs to the ABI type of their fields */
+interface StructFields {
+  [fieldName: string]: ABIType | StructFields;
+}
+```
+
+### Storage Interfaces
+
+These interfaces properly describe how app storage is access within the contract
+
+```ts
+/** Describes a single key in app storage */
+interface StorageKey {
+  /** Description of what this storage key holds */
+  desc?: string;
+  /** The type of the key */
+  keyType: ABIType | AVMBytes | StructName;
+  /** The type of the value */
+  valueType: ABIType | AVMBytes | StructName;
+  /** The bytes of the key encoded as base64 */
+  key: string;
+}
+
+/** Describes a mapping of key-value pairs in storage */
+interface StorageMap {
+  /** Description of what the key-value pairs in this mapping hold */
+  desc?: string;
+  /** The type of the keys in the map */
+  keyType: ABIType | AVMBytes | StructName;
+  /** The type of the values in the map */
+  valueType: ABIType | AVMBytes | StructName;
+  /** The base64-encoded prefix of the map keys*/
+  prefix?: string;
+}
+```
+
+### SourceInfo Interface
+
+This interface gives clients more information about the contract's source code.
+
+```ts
+interface SourceInfo {
```

Comment:
tealscript also seems to have a `source` property

@daniel-makerx is currently looking at how to represent source maps from high level language, probably worth both of you having a chat and seeing what makes sense?

Reply:
Looks like the `source` property points to the source line in a high level language. It would only make sense to include that in ARC-56 if the source path and contents was included as well. But even that can get tricky as there can be multiple files involved.

If the primary goal of SourceInfo is to provide some basic error/debugging info for TEAL source, then I think what is in the spec currently is sufficient.

With regards to the source maps we are looking at for Puya, they will include more information than just PC -> line mappings, ideally enough so that resolving variables and their types to AVM stack values is possible, as well as preserving call stack information.

Reply:
Agreed.

Reply:
@joe-p thoughts on what we do to resolve this? The typed client TypeScript output doesn't like the extra source property, if we want to include it we should put it in the spec here as an optional?

Reply:
Ah right yeah... it is really only useful for the developer which will have the full source code available to them. I think adding `source?: string` makes the most sense, with the values the file and line no ie. `somefile.py:1337`

Reply:
I'd also say that that info is mainly useful for breakpoint debugging of HLL source? In which case the work that Daniel and Al are doing will set the "how" (and it's not via ARC-56). Loading that level of debugging info doesn't feel that useful and adds significant weight to the file, should we leave it out?

Reply:
Breakpoint debugging is useful when you are running tests, but there might be unexpected errors than are encountered when integrating into a frontend or in production in which case you'll need to work backwards from the PC to high-level source

Reply:
Exactly my thinking. Otherwise the arc-56 file is huge. I mean it is anyway with the teal source info, but even more so with HLL source...

Reply:
Taking it a step further... perhaps for ARC56 it makes sense to only include SoureInfo objects that have an error message

Reply:
And also omit TEAL line mapping

Reply:
Assuming we are happy with just doing program counter with the implication that means an extra HTTP request sometimes, then yep I think that makes sense.

Reply:
When would you need to do the extra HTTP request?

Reply:
To grab the deployed byte code (although on reflection, we can stash that away like we do for source maps today so on reflection it's all good actually)

Reply:
To be clear I don't mean removing source entirely, just the TEAL lines from the mapping. So in the end you just have a mapping of PCs to error messages. 19c62692572f111edd93cf673863ac64bd33c84e

Reply:
Yep understood and I concur :)